contracts/UniswapV3Staker.sol

```diff
@@ -253,56 +235,74 @@ contract UniswapV3Staker is
-            by secondsX96 to get reward.
-        * totalRewardsUnclaimed is decremented by reward. totalSecondsClaimed
-            is incremented by seconds.
-        */
-
-        // TODO: check for overflows and integer types
-        // uint160 secondsX96 = FullMath.mulDiv(secondsPerLiquidityInStakingPeriodX128, , denominator);
-        //     SafeMath.mul(secondsPerLiquidityInStakingPeriodX128, liquidity);
-
-        incentives[incentiveId].totalSecondsClaimedX128 += secondsInPeriodX128;
+            uint160(
+                SafeMath.mul(
+                    secondsPerLiquidityInsideX128 -
+                        stakes[params.tokenId][incentiveId]
+                            .secondsPerLiquidityInitialX128,
+                    liquidity
+                )
+            );
 
+        // TODO: double-check for overflow risk here
         uint160 totalSecondsUnclaimedX128 =
-            uint32(Math.max(params.endTime, block.timestamp)) -
-                params.startTime -
-                incentives[incentiveId].totalSecondsClaimedX128;
-
-        // This is probably wrong
-        uint160 rewardRate =
             uint160(
-                SafeMath.div(
-                    incentives[incentiveId].totalRewardUnclaimed,
-                    totalSecondsUnclaimedX128
-                )
+                SafeMath.mul(
+                    Math.max(params.endTime, block.timestamp) -
+                        params.startTime,
+                    FixedPoint128.Q128
+                ) - incentives[incentiveId].totalSecondsClaimedX128
+            );
+
+        // TODO: Make sure this truncates and not rounds up
+        uint256 rewardRate =
+            FullMath.mulDiv(
+                incentives[incentiveId].totalRewardUnclaimed,
+                FixedPoint128.Q128,
+                totalSecondsUnclaimedX128
             );
 
-        uint256 reward = SafeMath.mul(secondsInPeriodX128, rewardRate);
+        // TODO: make sure casting is ok here
+        uint128 reward =
+            uint128(
+                FullMath.mulDiv(
+                    secondsInPeriodX128,
+                    rewardRate,
+                    FixedPoint128.Q128
+                )
+            );
 
-        // TODO: Before release: wrap this in try-catch properly
-        // try {
-        // TODO: incentive.rewardToken or rewardToken?
+        incentives[incentiveId].totalSecondsClaimedX128 += secondsInPeriodX128;
 
-        IERC20Minimal(incentives[incentiveId].rewardToken).transfer(
-            params.to,
-            reward
+        // TODO: is SafeMath necessary here? Could we do just a subtraction?
+        incentives[incentiveId].totalRewardUnclaimed = uint128(
+            SafeMath.sub(incentives[incentiveId].totalRewardUnclaimed, reward)
         );
-        // } catch {}
+
+        /* TODO: This will be fixed in https://github.com/omarish/uniswap-v3-staker/issues/38
+        so that collecting and unstaking are two separate functions */
+        try
```

Comment:
might be neater to separate transfer out completely into its own function. (w/ multicall can be done in one tx) Convo in https://github.com/omarish/uniswap-v3-staker/issues/54 @femi